qiskit/providers/experiment/experiment_data.py

```diff
@@ -0,0 +1,674 @@
+            notes: Freeform notes about the experiment.
+
+        Raises:
+            ExperimentError: If an input argument is invalid.
+        """
+        metadata = metadata or {}
+        self._source = metadata.pop(
+            "_source",
+            {"class": f"{self.__class__.__module__}.{self.__class__.__name__}",
+             "data_version": self._metadata_version,
+             "_qiskit_version": __qiskit_version__})
+
+        self._service = None
+        self._backend = backend
+        self.auto_save = False
+        try:
+            self._service = backend.provider().service('experiment')
+            self.auto_save = self._service.option['auto_save']
+        except Exception:  # pylint: disable=broad-except
+            pass
+
+        self._id = experiment_id or str(uuid.uuid4())
+        self._type = experiment_type
+        self._tags = tags or []
+        self._share_level = share_level
+        self._notes = notes or ""
+        self._metadata = copy.deepcopy(metadata)
+
+        job_ids = job_ids or []
+        self._jobs = OrderedDict((k, None) for k in job_ids)
+        self._job_futures = []
+        self._errors = []
+
+        self._data_queue = queue.Queue()
+        self._data = []
+
+        figure_names = figure_names or []
+        self._figures = dict.fromkeys(figure_names)
+        self._figures_queue = queue.Queue()
+        self._figure_names = figure_names  # Only used by add_figure.
+
+        self._analysis_results_queue = queue.Queue()
+        self._analysis_results = []
+
+        self._created_in_db = False
+
+    def add_data(
+            self,
+            data: Union[Result, List[Result], Job, List[Job], Dict, List[Dict]],
+            post_processing_callback: Optional[Callable] = None
+    ):
+        """Add experiment data.
+
+        Args:
+            data: Experiment data to add.
+                Several types are accepted for convenience:
+
+                    * Result: Add data from this ``Result`` object.
+                    * List[Result]: Add data from the ``Result`` objects.
+                    * Job: Add data from the job result.
+                    * List[Job]: Add data from the job results.
+                    * Dict: Add this data.
+                    * List[Dict]: Add this list of data.
+            post_processing_callback: Callback function invoked when all pending
+                jobs finish. This ``ExperimentData`` object is the only argument
+                to be passed to the callback function.
+        """
+        if isinstance(data, dict):
+            self._add_single_data(data)
+        elif isinstance(data, Result):
+            self._add_result_data(data)
+        elif isinstance(data, (Job, BaseJob)):
+            self._jobs[data.job_id()] = data
+            self._job_futures.append(
+                (data, self._executor.submit(self._wait_for_job, data, post_processing_callback)))
```

Comment:
Can you add KeyboardInterrupt exception to cancel the job? I may want to stop waiting a job when I find mistake in the experiment setting, but now the job is wrapped by experiment cannot easy access to the cancel method.